c/tskit/trees.c

```diff
@@ -4612,76 +4642,347 @@ norm_kc_vectors(kc_vectors *self, kc_vectors *other, double lambda)
     return sqrt(distance_sum);
 }
 
+static int
+check_kc_distance_tree_inputs(tsk_tree_t *self)
+{
+    tsk_id_t u, num_nodes, num_samples, left_child;
+    int ret, i;
+    ret = 0;
+
+    if (tsk_tree_get_num_roots(self) != 1) {
+        ret = TSK_ERR_MULTIPLE_ROOTS;
+        goto out;
+    }
+    if (!tsk_tree_has_sample_lists(self)) {
+        ret = TSK_ERR_UNSUPPORTED_OPERATION;
+        goto out;
+    }
+
+    num_samples = (tsk_id_t) self->tree_sequence->num_samples;
```

Comment:
Ah, so this is an issue. We can't go through the n samples for each tree to check if any of them are internal (an external sample in one tree could be internal in the next). We'll just have to relax the requirement, and raise an issue to make sure we follow up and document what happens when we run KC on trees with internal samples. (It all probably works just fine, we'd just need to document what happens and put in a few test cases to make sure it's doing something sensible).

Reply:
I get that it would be nice to allow it on internal samples but wouldn't this be the same behavior as in the single tree case? The whole thing fails if any individual comparisons would fail. I guess to fix it long term the question is does KC operate on samples or leaves?

Reply:
I think we need to think of KC as working on samples rather than leaves. At the moment, we're assuming a 1-1 correspondence between the two, but I don't think it's hard to tweak the definition so that it makes sense for arbitrary samples (might even do it out of the box).

Anyway, for now, let's just get rid of this check in the TS version because we don't want to incur this O(n) cost for each tree, write a test to make sure things don't blow up when we have internal samples, and then follow up with an issue to make sure that we define KC properly on arbitrary samples.

Reply:
Ah ok, sounds good!